working/0546-patterns/patterns-feature-specification.md

```diff
@@ -2159,17 +2209,233 @@ To match a pattern `p` against a value `v`:
+it only gets invoked once even when used by multiple patterns.
+
+It works like this:
+
+1.  At compile time, after type checking has completed, we associate an
+    *invocation key* with every member call or record field access potentially
+    made by each pattern.
+
+2.  At runtime, whenever the runtime semantics say to call a member or access a
+    record field, if a previous call or access with that same invocation key has
+    already been evaluated, we reuse the result.
+
+3.  Otherwise, we invoke the member or field access now and associate the result
+    with that invocation key for future calls.
+
+Let an *invocation key* comprise:
+
+*   A possibly absent parent invocation key.
+*   A possibly absent extension type. If the invocation represents an extension
+    member call, this tracks the type the call was resolved to.
+*   A member name.
+*   A possibly empty list of argument constant values.
+
+Two invocation keys are equivalent if and only if all of these are true:
+
+*   They both have parent invocation keys and the keys are equivalent or
+    neither of them have parent invocation keys.
+*   The extension types refer to the same type or are both absent.
+*   The member names are equivalent.
+*   The argument lists have the same length and all corresponding pairs of
+    argument constant values are identical.
+
+*In other words, they're equal if all of their fields are equal in the obvious
+ways.*
+
+The notation `parent : (name, args)` creates an invocation key with parent
+`parent`, no extension type `type`, member name `name`, and argument list
+`args`. The notation `parent : type(name, args)` creates an invocation key with
+parent `parent`, extension type `type`, member name `name`, and argument list
+`args`.
+
+Given a set of patterns `s` matching a value expression `v`, we bind an
+invocation key to each member invocation and record field access in `s` like so:
+
+1.  Let `i` be an invocation key with no parent, no extension type, named
+    `this`, with an empty argument list. *This is the root node of the
+    invocation key tree and represents the matched value itself.*
+
+2.  For each pattern `p` in `s` with parent invocation `i`, bind invocation keys
+    to it and its subpatterns using the following procedure:
+
+To bind invocation keys in a pattern `p` using parent invocation `i`:
+
+*   **Logical-or** or **logical-and**:
+
+    1.  Bind invocations in the left and right subpatterns using target `i`.
+
+*   **Relational**:
+
+    1.  Resolve the relational operator to a member.
+
+    2.  If it resolves to an instance method, then bind `i : (op, [arg])` to
+        the operator method invocation where `op` is the name of the operator
+        and `arg` is the right operand value.
+
+    3.  Else it resolves to an extension method. Bind `i : type(op, [arg])` to
+        the operator method invocation where `type` is the type the extension
```

Comment:
Not a type. I don't remember if we have a name for an "extension instantiation", but that phrase can be used.

Reply:
Fixed.